### PR TITLE
Fix tenant selection for username in reCaptcha

### PR DESCRIPTION
**Pull request by ashans**
### Proposed changes in this pull request

Fix reCaptcha related configurations getting from super tenant when tenant user tries to login to myaccount or any other service provider. Changed the tenant retrieving from username to tenant retrieving from Authentication context.

### CODE REVIEW

File: components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java [hunk 1]
```diff
@@ -298,4 +306,28 @@ public Properties getDefaultPropertyValues(String tenantDomain) throws IdentityG
...
+
+    /**
+     * Get tenant from servlet request or username.
+     *
+     * @param servletRequest    Servlet request.
+     * @param userName          Username.
+     * @return                  Derived tenant domain.
+     */
+    private String getTenant(ServletRequest servletRequest, String userName) {
+        if (IdentityTenantUtil.isTenantedSessionsEnabled() || IdentityTenantUtil.isTenantQualifiedUrlsEnabled()) {
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java:318]: We need new line after the method name

File: components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java [hunk 2]
```diff
@@ -298,4 +306,28 @@ public Properties getDefaultPropertyValues(String tenantDomain) throws IdentityG
...
     }
+
+    /**
+     * Get tenant from servlet request or username.
+     *
+     * @param servletRequest    Servlet request.
+     * @param userName          Username.
+     * @return                  Derived tenant domain.
+     */
+    private String getTenant(ServletRequest servletRequest, String userName) {
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java:317]:
```suggestion
    private String getTenant(ServletRequest servletRequest, String username) {
```

File: components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java (hunk 1 above)
Comment (ashans) [reply components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java:319]: Fixed in 10462a5

File: components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java (hunk 2 above)
Comment (ashans) [reply components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java]: Fixed in 10462a5

File: components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java [hunk 3]
```diff
@@ -298,4 +306,29 @@ public Properties getDefaultPropertyValues(String tenantDomain) throws IdentityG
...
+
+            AuthenticationContext context = FrameworkUtils.getAuthenticationContextFromCache(sessionDataKey);
+            if (context == null) {
+                return null;
+            }
+            return context.getUserTenantDomain();
+        } else {
+            return MultitenantUtils.getTenantDomain(username);
+        }
+    }
```
Comment (omindu) [inline components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java:333]:
```suggestion
    private String getTenant(ServletRequest servletRequest, String username) {

        if (IdentityTenantUtil.isTenantedSessionsEnabled() || IdentityTenantUtil.isTenantQualifiedUrlsEnabled()) {
            return context.getUserTenantDomain();
        } else {
            return MultitenantUtils.getTenantDomain(username);
        }
    }
```

File: components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java [hunk 4]
```diff
@@ -80,7 +83,11 @@ public boolean canHandle(ServletRequest servletRequest, ServletResponse servletR
             return false;
         }
```
Comment (omindu) [inline components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java:85]:
```suggestion
            String sessionDataKey = servletRequest.getParameter(FrameworkUtils.SESSION_DATA_KEY);
            if (sessionDataKey == null) {
                return false;
            }
            AuthenticationContext context = FrameworkUtils.getAuthenticationContextFromCache(sessionDataKey);
            if (context == null) {
                return false;
            }
```

File: components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java [hunk 5]
```diff
@@ -80,7 +83,11 @@ public boolean canHandle(ServletRequest servletRequest, ServletResponse servletR
...
 
-        String tenantDomain = MultitenantUtils.getTenantDomain(userName);
+        String tenantDomain = getTenant(servletRequest, userName);
```
Comment (omindu) [inline components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java:86]:
```suggestion
        String tenantDomain = getTenant(context, userName);
```

File: components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java (hunk 3 above)
Comment (ashans) [reply components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java]: Fixed in 2817200

File: components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java (hunk 4 above)
Comment (ashans) [reply components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java:85]: Fixed in 2817200

File: components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java (hunk 5 above)
Comment (ashans) [reply components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java]: Fixed in 2817200

File: components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java [hunk 6]
```diff
@@ -298,4 +317,20 @@ public Properties getDefaultPropertyValues(String tenantDomain) throws IdentityG
         failedRedirectUrls.add(ON_FAIL_REDIRECT_URL);
         return failedRedirectUrls;
     }
+
+    /**
+     * Get tenant from servlet request or username.
```
Comment (omindu) [inline components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java:322]:
```suggestion
     * Get tenant from authentication context or username.
```

Comment (ashans) [reply components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java]: Fixed in bfff90b

**Review by omindu — APPROVED** [review]

**Pull request merged**